de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java

```diff
@@ -0,0 +1,238 @@
+ * #L%
+ */
+
+/**
+ * A helper class for {@link InventoryImportProcess} that performs the "dirty" but efficient SQL updates on the {@link I_I_Inventory} table.
+ * Those updates complements the data from existing metasfresh records and flag those import records that can't yet be imported.
+ *
+ * @author metas-dev <dev@metasfresh.com>
+ *
+ */
+@UtilityClass
+public class MInventoryImportTableSqlUpdater
+{
+	private static final transient Logger logger = LogManager.getLogger(MInventoryImportTableSqlUpdater.class);
+
+	@Builder(buildMethodName = "updateIInventory")
+	private void updateInventoryImportTable(@NonNull final String whereClause)
+	{
+		dbUpdateLocatorDimensions(whereClause);
+		dbUpdateWarehouse(whereClause);
+		dbUpdateCreateLocators(whereClause);
+		dbUpdateProducts(whereClause);
+		dbUpdateSubProducer(whereClause);
+
+		dbUpdateErrorMessages(whereClause);
+	}
+
+	private void dbUpdateLocatorDimensions(@NonNull final String whereClause)
+	{
+		// Set M_Warehouse_ID
+		StringBuilder sql = new StringBuilder("UPDATE I_Inventory i ")
+				.append("SET warehouseValue = dimensions.warehouseValue, X = dimensions.locatorX , Y = dimensions.locatorY, Z = dimensions.locatorZ, X1 = dimensions.locatorX1 ")
+				.append("FROM (SELECT I_Inventory_ID, d.warehouseValue, d.locatorValue, d.locatorX, d.locatorY, d.locatorZ, d.locatorX1 ")
+				.append("	FROM I_Inventory as inv")
+				.append("	JOIN extractLocatorDimensions(inv.locatorvalue) as d on d.locatorvalue=inv.locatorvalue")
+				.append(") AS dimensions ")
+				.append("WHERE I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Warehouse from Locator =" + no);
+	}
+
+	private void dbUpdateWarehouse(@NonNull final String whereClause)
+	{
+		// Set M_Warehouse_ID
+		final StringBuilder sql = new StringBuilder("UPDATE I_Inventory i ")
+				.append("SET M_Warehouse_ID=(SELECT M_Warehouse_ID FROM M_Warehouse w WHERE i.warehouseValue=w.value) ")
+				.append("WHERE M_Warehouse_ID IS NULL ")
+				.append("AND I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Warehouse =" + no);
+	}
+
+	private void dbUpdateCreateLocators(@NonNull final String whereClause)
+	{
+		dbUpdateLocators(whereClause);
+		dbCreateLocators();
+	}
+
+	private void dbUpdateLocators(@NonNull final String whereClause)
+	{
+		StringBuilder sql = new StringBuilder("UPDATE I_Inventory i ")
+				.append("SET M_Locator_ID=(SELECT MAX(M_Locator_ID) FROM M_Locator l ")
+				.append("WHERE i.LocatorValue=l.Value AND i.AD_Client_ID=l.AD_Client_ID) ")
+				.append("WHERE M_Locator_ID IS NULL AND LocatorValue IS NOT NULL ")
+				.append("AND I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Locator from Value =" + no);
+	}
+
+	private void dbCreateLocators()
```

Comment:
what if on 2 different I_Inventory records you have same locator (i.e. same x,y,z,x1 etc) ?

Reply:
is fine, same locator, probably different product